apps/train_loc/lib/train_loc/manager/bulk_event.ex

```diff
@@ -0,0 +1,116 @@
+                    },
+                    "required" => ["date", "results"]
+                  },
+                  "results" => %{
+                    "type" => "object",
+                    "additionalProperties" => %{
+                      "type" => "object",
+                      "properties" => %{
+                        "Heading" => %{"type" => "number"},
+                        "Latitude" => %{"type" => "number"},
+                        "Longitude" => %{"type" => "number"},
+                        "Source" => %{"type" => "string"},
+                        "Speed" => %{"type" => "integer"},
+                        "TripID" => %{"type" => "string"},
+                        "Update Time" => %{"type" => "string", "format" => "date-time"},
+                        "VehicleID" => %{"type" => "integer"}
+                      },
+                      "required" => [
+                        "VehicleID",
+                        "Heading",
+                        "Latitude",
+                        "Longitude",
+                        "Speed",
+                        "TripID",
+                        "Update Time"
+                      ]
+                    }
+                  }
+                },
+                "required" => ["processResults", "results"]
+              }
+            },
+            "required" => ["data", "path"]
+          }
+          |> ExJsonSchema.Schema.resolve()
+
+  @spec valid_json?(map) :: boolean
+  def valid_json?(event) when is_map(event),
+    do: ExJsonSchema.Validator.valid?(@schema, event)
+
+  @spec parse(map | String.t()) ::
+          {:ok, [NewVehicle.t()]}
+          | {:error, :invalid_event, map}
+          | {:error, Jason.DecodeError.t()}
+  def parse(event) when is_map(event) do
+    if valid_json?(event) do
+      {:ok,
+       event["data"]["results"]
+       |> Map.values()
+       |> Enum.flat_map(&parse_raw_vehicle/1)}
+    else
+      {:error, :invalid_event, event}
+    end
+  end
+
+  def parse(event) when is_binary(event) do
+    case Jason.decode(event, strings: :copy) do
+      {:ok, event} ->
+        parse(event)
+
+      {:error, %Jason.DecodeError{} = error} ->
+        {:error, error}
+    end
+  end
+
+  @spec parse_raw_vehicle(map) :: [TrainLoc.Vehicles.NewVehicle.t()]
+  defp parse_raw_vehicle(%{
+         "Heading" => heading,
+         "Latitude" => latitude,
+         "Longitude" => longitude,
+         "Speed" => speed,
+         "TripID" => trip_id,
+         "Update Time" => update_time,
+         "VehicleID" => vehicle_id
+       }) do
+    case DateTime.from_iso8601(update_time) do
+      {:ok, timestamp, _} ->
+        [
+          %NewVehicle{
+            vehicle_id: vehicle_id,
+            timestamp: timestamp,
+            heading: heading,
+            speed: speed,
+            latitude: latitude,
+            longitude: longitude,
+            trip: trip_id
+          }
+        ]
+
+      _error ->
```

Comment:
question: should this log the invalid timestamp?

Reply:
Probably.